presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -653,9 +656,29 @@ private AWSCredentialsProvider getAwsCredentialsProvider(URI uri, Configuration
             return new InstanceProfileCredentialsProvider();
         }
 
+        String customCredentialsProviderClass = conf.get(S3_CREDENTIALS_PROVIDER);
+        if (customCredentialsProviderClass != null && !customCredentialsProviderClass.isEmpty()) {
+            return getCustomAWSCredentialsProvider(uri, conf, customCredentialsProviderClass);
+        }
+
         throw new RuntimeException("S3 credentials not configured");
     }
 
+    private AWSCredentialsProvider getCustomAWSCredentialsProvider(URI uri, Configuration conf, String customCredentialsProviderClass)
+    {
+        try {
+            AWSCredentialsProvider credentials = (AWSCredentialsProvider) Class
+                    .forName(customCredentialsProviderClass)
+                    .getDeclaredConstructor(URI.class, Configuration.class)
```

Comment:
This should use `getConstructor` so that it only returns public constructors (`getDeclaredXXX` can return private)